### ISSUE TITLE
There is no way to define a service without a dockerfile or image, right? How much work would a fork / PR for this be roughly?

### ISSUE THREAD

**Issue by steffres** [docker/compose#10758]
Hi,

### TL;DR:
- It's not possible to define a service within a compose.yml that does not use a dockerfile or image, right?
- How much work would you estimate it to be, if I chose to fork compose and implement such a feature?
- Could this be of use upstream again for compose? Or is it scope creeping?
- Would there be other tools that serve our purpose better?

### Motivation:

I work in academia / linguistics, and we use docker and compose heavily. Besides the main reason for docker (easy encapsulation), we also use it more and more for reliable reproduction of data processing pipelines for scientific quality assurance. I.e. we define services handling data within a compose.yml which is versioned with git. This way most data processing can be persisted and made reproducable by just launching a service with a compose.yml at a certain git commit in time while also having the mounted data repos and source code equally commited (and pointed to with git submodules). 

E.g.:
```
compose.yml 1, defining such a pipeline, at some commit, also pointing to submodules:

[data repo 1, at some commit] 
|
mounted as volume, and used as input for code
|
V
[code repo 1, with micro compose.yml and dockerfile, at some commit] 
|
mounted as volume, and produced as output by code
|
V
[data repo 2, at some commit]
```

This setup works great. However it has the disadvantage of not being as modular as we would like it to be. Because we would like to use any compoment of this chain in arbitrary other chains as well. E.g. the same input data as before but with another application: `code repo 2`, defined as a whole in `compose.yml 2`.:
```
compose.yml 2:

[data repo 1]
|
V
[code repo 2]
```

Code repos can already be individually defined with a self-contained compose.yml that can be reused in containing compose.yml with `extends`, **but not data repos as they don't have a dockerfile or image. This hinders our aim for increased modularity, where we want to arbitrarily chain data and processing together and persist this whole pipeline for reproducability**

Such a "data service" (i.e. compose service withou dockerfile or image) can't be easily done without dirty hacks in vanilla compose, right?

### Suggestion

For our purpose it would be great if we could define a static "data service" within a compose.yml that only serves the purpose of providing indexable metadata (what kind of data mostly, what compability with what tools) so that it can be reused in arbitrary containing compose.yml with `extends`.

I consider allocating my time into such feature. But before doing so, I kindly ask:
- How much work would it be to fork compose and implement such a feature? 
- Is such a feature desirable by compose or is it scope creeping?
- or are there other tools that serve our purpose better (I didn't find any)?

Cheers,
Stefan

**Comment by laurazard** [comment]
Hiya @steffres,

Maybe I'm misreading since I don't fully understand your usecase and what constraints you face, but from a cursory glance there are a few ways of achieving what you like with the functionality that's available in Compose today. I can't tell from your description whether the data you want to codify into your Compose files is mere data (as in, files in a filesystem) or a service which provides such data (in which case, I don't see why you wouldn't be able to implement a 
> static "data service" within a compose.yml that only serves the purpose of providing indexable metadata

yourself, with your own protocol (I'm not sure Compose should define a protocol for data sharing and advertising compatibility with any tools, as that kind of protocol/implementation is usually best left to the user to implement for themselves since it's highly dependent on their specific usecase). If your data is just that, data and files on a filesystem, there are other approaches you can take – a common pattern is to define a volume in your Compose file, and use a "startup service" (which your processing service depends on) whose job is to, on startup, populate the volume that's going to contain the data for your other service with the data you like. In practice, that service would be your "data service" which you could override from other Compose files (set environment vars, etc.) to control which data is downloaded. Regarding this pattern, there are proposals in the Compose Spec currently (https://github.com/compose-spec/compose-spec/issues/291 comes to mind) that would make this sort of workflow more ergonomic. Other approaches users sometimes take is just to store/work with data *as* images (although that can be quite inneficient).

A quick, relevant aside is that Compose (this repo) conforms to the [Compose Specification](https://github.com/compose-spec/compose-spec) and this type of discussion is generally preferred over there, as it impacts not just Compose functionality/bugs but overall Spec design and expected behaviours. If you like, you can create an RFC/Proposal over https://github.com/compose-spec/compose-spec/issues describing your usecase/thoughts and/or link it back here.

**Comment by steffres** [comment]
Hi Laura,

thank you for your reply and inputs!

On:

**custom metada**

> I'm not sure Compose should define a protocol for data sharing and advertising compatibility with any tools, as that kind of protocol/implementation is usually best left to the user to implement for themselves since it's highly dependent on their specific usecase

Precisely. We plan to reuse whatever container / compose practices are established, and then, on top of this, we will use compose's `x-` keys where we inject our custom meta-data for our service to discover and link things together that should be link-able.

**startup service**

Thanks for the suggestion of the "startup service" and the link, I will look into this and experiment with it. Might be helpful. I will update here.

**image as data**

On the suggestion of building data images, yeah, we considered that but besides this being a hackish work-around, it also defeats the purpose of having an easy quick look into the data on a plain filesystem. Also data will be produced by our services too and persisted into git repos, for which we would need to mount the image and read it. While technically possible, it's probably overall too cumbersome.

**compose spec issues**

Thanks for the link to the spec repo, I'll open an issue there too and link it.

**Comment by steffres** [comment]
**startup service**

I could create a dummy setup. 

folder:
```
├── compose.yml
├── service_code
│   ├── compose.yml
└── service_data
    ├── compose.yml
    └── data.json
```
where 

`compose.yml`
```
services:
    service_code:
        extends:
            file: ./service_code/compose.yml
            service: service_code
        volumes_from:
            - service_data

    service_data:
        extends:
            file: ./service_data/compose.yml
            service: service_data
```

`service_code/compose.yml`
```
services:
    service_code:
        image: bash:latest
```

`service_data/compose.yml`
```
services:
    service_data:
        image: busybox:latest
        volumes:
            - ./:/data
```

If I start the main service and enter the it, I can see `data.json` in a folder mounted at `/data`.
```
steff-vm@base-vm:~/exp/docker_exp$ docker-compose run service_code
Creating docker_exp_service_data_1 ... done
Creating docker_exp_service_code_run ... done
bash-5.2# ls /data/
compose.yml  data.json
```

So, yes, that works. But it's not satisfying as I need an image just to serve a data volume. I'd really like to leave that data image out, as we want to provide our setup to downstream users and make it as easily adaptable as possible.

If there are other ideas, I'm all open please.

**Comment by ipedrazas** [comment]
Hello @steffres 

I can tell you how I handle these situations and see if that helps with your use case.

If your data is in a git repo, the easiest way to have that data in your Compose is to have a service that checks out that data into a volume. I don't like this approach because your compose is a lightweight wrapper on git (credentials, size, etc.), so, you need to make sure that whoever uses that service, has all the data/access necessary.

What I usually do is package my data as OCI images. This turns data sets into content-addressable data sets. What we do is swap the git sha as the ID of the data set with the image digest. Both of them are immutable, but an OCI image is content addressable (can be pulled from different places).

For example, this is one of the ways I run websites. I have an nginx as my webserver, and the data is packaged as an OCI image that is loaded before the Nginx starts:

```
services:
  webiste:
    image: nginx:latest
    volumes:
      - web_data:/usr/share/nginx/html
    ports:
      - 8000:80
    depends_on:
       data:
          condition: service_completed_successfully

  data:
    image: ipedrazas/ktools:latest
    user: root
    command: ["oras","pull", "registry-1.docker.io/ipedrazas/blog-oci:latest", "-o", "/usr/share/nginx/html"]
    volumes:
      - web_data:/usr/share/nginx/html

volumes:
  web_data:
```
  
The `data` service just pulls the data. I think we can agree that it would be nicer if we had something like this instead: 


```
services:
  webiste:
    image: nginx:latest
    volumes:
      - type: oci
         source: registry-1.docker.io/ipedrazas/blog-oci:latest
         target: /usr/share/nginx/html
    ports:
      - 8000:80
```

But the real gain is that we move data sets management into CI that produces data artifacts (yes, you can sign them also, in case you're interested in providing attestations). It is a bit more work on the data-producing side, but it provides more control AND it is a nicer experience when consuming data :)

I have a little repo where I explain these data patterns with OCI: https://github.com/ipedrazas/oci-patterns

For example, this is how I run my gitea server: https://github.com/ipedrazas/oci-patterns/tree/main/oci-and-compose/gitea and yes, all my backups and restores are now OCI based. Creating a backup is doing a `docker build` and restoring it is a `docker run`. 

You can use a `Dockerfile` to define how your data set is created, for example, you can do something like this to create an OCI image with the contents of the current directory:

```
FROM scratch

COPY . /
```

And create the image by doing:

```
docker build -o type=image,oci-mediatypes=true,name=ipedrazas/oci-data:latest .
```

Or, in the case of the data being in git... you could do this:

```
FROM alpine/git as builder
ARG BRANCH
ENV BRANCH?="main"

RUN <<EOF
  git clone https://github.com/compose-spec/compose-spec.git
  cd compose-spec
  git checkout ${BRANCH}
EOF

FROM scratch
COPY --from=builder /git/compose-spec /data
```

and build it like: `docker build -e build-arg BRANCH=golang -o type=image,oci-mediatypes=true,name=ipedrazas/compose-spec-oci:latest . --push`


Happy to chat more about this if you're interested, you can find me in the [Community Docker Slack ](dockercommunity.slack.com) or [Twitter](https://twitter.com/ipedrazas).

**Comment by ndeloof** [comment]
@ipedrazas this can be implemented using a custom volume driver installed on docker engine to create local volume from an OCI artifact.

**Comment by steffres** [comment]
Hey @ipedrazas 

thanks for your extensive reply and your approach. I consider packing data into images as well, for all the advantages you have listed. Foremost this approach shines if the ratio of consuming data versus producing data is more on the consuming side.

However in our case we also produce a lot of data in short intervals, where simple git or dvc is a smoother straight forward UX for our scholars. Our highest priority is to make it as easy as possible to play around with the data, say someone downloads the repo / dvc or whatever and fiddles with in their applications. If the data were encapsulated in a docker iamge, there's a step in between, where the data would need to be mounted and then persisted into an image again. 

Our overall heavily dockerized approach should just sprinkle some reliable execution logic on top of existing data sets and tooling, and not replace some established practices.

Due to these reasons, generating data images is just a bit not smooth enough.

**Comment by steffres** [comment]
> @ipedrazas this can be implemented using a custom volume driver installed on docker engine to create local volume from an OCI artifact.

Thanks for your input. May I ask you to point me to reference material for this please? When I google "google custom volume driver" or something similiar I only find downstream application / configurations of tools as docker compose or kubernetes, but nothing that goes under the hood as I would need it.

**Comment by ndeloof** [comment]
@steffres see https://docs.docker.com/engine/reference/commandline/plugin/

**Comment by steffres** [comment]
great, thank you!

**Comment by steffres** [comment]
@ndeloof may I ask you something slightly related please? As I see your activities in your profile are quite close to all of this, do you happen to know the answer to this question I had posted over in compose-spec: https://github.com/compose-spec/compose-spec/issues/379 ?

**Comment by steffres** [comment]
Alright. Overall, I see three options:

- provide data as docker images
- write custom volume driver 
- don't use docker compose for data 

We will try to go with not using docker compose for data, but will create a meta data for both data and services (which describes with kind of data can be ingested by which kind of service). And that will be a separate config file that lies in the repos and that tell our discoverability services what to do. If some dataset is compatible with some service, then this data set can be mounted as classical volume. 

This approach is overall slightly less consistent (because of different handling for data and for services), but much more feasible and based on common practices.

Should we not be happy with this approach we might try the custom volume driver (more likely) or even data images (less likely).

Thanks again!
